[runtimes][PAC] Harden unwinding when possible

ojhunt opened the pull request:
This hardens the unwinding logic and datastructures on systems
that support pointer authentication.

The approach taken to hardening is to harden the schemas of as many
high value fields in the myriad structs as possible, and then also
explicitly qualify local variables referencing privileged or security
critical values.

This does introduce ABI linkage between libcxx, libcxxabi, and
libunwind but those are in principle separate from the OS itself
so we've kept the schema definitions in the library specific headers
rather than ptrauth.h

asl commented:
@kovdan01 @atrosinenko Please take a look

ojhunt commented:
Updating formatting before review - had discussed with Louis and he expressed a preference for some of it, but this llvm.org style bot complains many other cases (likely a local config issue when I was trying to cleanup the downstream code.

So I've updated with a direct clang-format fix.

ojhunt commented on libcxxabi/include/__cxxabi_config.h:
```diff
@@ -32,22 +32,23 @@
-  #define _LIBCXXABI_FUNC_VIS __declspec(dllimport)
-  #define _LIBCXXABI_TYPE_VIS __declspec(dllimport)
- #endif
+#  if defined(_LIBCXXABI_DISABLE_VISIBILITY_ANNOTATIONS) ||                                                            \
```
I'm not sure how/why/when I changed this formatting because I'd swear I went through trying to remove them all

ojhunt replied:
Oh, I bet I auto-ran clang-format after the bots complained or reflexively pre-push to avoid the bots complain. Sighhhhh

ojhunt commented on libunwind/src/UnwindRegistersSave.S:
```diff
@@ -807,7 +812,12 @@ DEFINE_LIBUNWIND_FUNCTION(__unw_getcontext)
 #endif
   mov    x0, #0                   // return UNW_ESUCCESS
+
+#if __has_feature(ptrauth_calls)
+  retab
+#else
```
This is what we have downstream, it's possible we've just developed an assumption that they're equivalent at this point.

That said I'm actually thinking this should also be around an `if __APPLE__` as well pending adoption of/a more generic gate on the ABI.

atrosinenko replied:
Same comment here (`autibsp` + `ret` instead of `retab`).

ojhunt commented on libcxxabi/src/cxa_personality.cpp:
```diff
@@ -557,7 +596,19 @@ set_registers(_Unwind_Exception* unwind_exception, _Unwind_Context* context,
   _Unwind_SetGR(context, __builtin_eh_return_data_regno(1),
                 static_cast<uintptr_t>(results.ttypeIndex));
+#if defined(__APPLE__) && __has_feature(ptrauth_qualifier)
+  auto stack_pointer = _Unwind_GetGR(context, UNW_REG_SP);
```
yeah, I spent a bunch of time trying to get the spellings more consistent but I really wasn't sure :-/

kovdan01 commented on libunwind/src/Registers.hpp:
```diff
@@ -1877,6 +1946,32 @@ inline Registers_arm64::Registers_arm64(const void *registers) {
+#endif
+}
+
+inline Registers_arm64::Registers_arm64(const Registers_arm64 &other) {
```
Declarations of these copy constructor and corresponding copy assignment operator are missing in the class body, so this basically does not seem to compile (at least I was unable to do that :) ). BTW, are these actually used somewhere?

ojhunt replied:
Hmmmm, I'll need to do some hunting to work out if this code is dead, or if I missed some declaration from

kovdan01 commented on libunwind/src/Registers.hpp:
```diff
@@ -93,6 +98,13 @@ class _LIBUNWIND_HIDDEN Registers_x86 {
 
+  typedef uint32_t reg_t;
+  typedef uint32_t link_reg_t;
+  void loadAndAuthenticateLinkRegister(reg_t srcLinkRegister,
```
Since `loadAndAuthenticateLinkRegister` is used in UnwindCursor.hpp as a member function of a `registers` object (having template registers class type), we should probably consider one of the following options.

1. Define `loadAndAuthenticateLinkRegister` for all the register classes. As far as I can see now, your patch contains definitions for x86, x86_64, ppc, arm64 and arm, but we also have mips, risc-v, ...
2. Only limit this function to arm64 and do not expose that to register classes for other architectures. In this case, you'll probably need to insert checks like `__has_feature(...)` in UnwindCursor.hpp and add special logic in these cases.
 
I personally prefer the 2nd option. It's not very nice, but IMHO this is better than adding a placeholder function for all the architectures.

kovdan01 commented on libunwind/src/Registers.hpp:
```diff
@@ -1845,10 +1871,53 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
+      __LIBUNWIND_PTRAUTH_RI_PDC("Registers_arm64::link_reg_t") link_reg_t;
+  void
+  loadAndAuthenticateLinkRegister(reg_t inplaceAuthedLinkRegister,
+                                  link_reg_t *referenceAuthedLinkRegister) {
```
Nit: probably it's worth using l-value reference instead of a pointer. We assume this being non-null, and actually each time you call this in UnwindCursor.hpp you are taking an address of a variable. L-value reference seems to be a perfect fit for this situation.

ojhunt replied:
I vastly prefer using pointers for out parameters as it makes it much clearer at the call site that it's an out parameter.

Ideally we'd just return the pointer but the problem is we can't have ptrauth-qualified return values and as a result it's a foot gun for an oracle.

Alternatively I could create a struct wrapper to force the authentication, but the problem there is that I think the design would want to construct a dependent ptrauth qualifier and we don't yet support that.

Or I guess we could just make a 
```cpp
template <class T> PointerAuthQualifierPointer = is_pointer<T> && (!ptrauth_available || ptrauth_schema_query_has_pointer_auth(T));
template < PointerAuthQualifierPointer T> struct PtrauthQualifiedPointer {
  T thePointer;
  ...
 }
```

To use for parameters and return values we consider important enough to require constant protection.

kovdan01 commented on libunwind/src/Registers.hpp:
```diff
@@ -1845,10 +1871,53 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
+  void
+  loadAndAuthenticateLinkRegister(reg_t inplaceAuthedLinkRegister,
+                                  link_reg_t *referenceAuthedLinkRegister) {
+#if __has_feature(ptrauth_calls)
```
Probably, you meant `ptrauth_returns`. I saw your discussion with @atrosinenko regarding similar situations in assembly files and I do get the point that these were assumed identical for your purposes in downstream. But it looks like that it's time to change this to `ptrauth_returns`. It's worth looking at other occurrences of `ptrauth_calls` as well - at least some of them are also probably misused and should be replaced with `ptrauth_returns`.

ojhunt replied:
We use a wide array of ptrauth_returns and ptrauth_calls, do we just want a single flag/define we should use?

I think the variation we see here is the outcome of changes over time, but it would seem reasonable to maybe just have a single mode flag at this point?

atrosinenko replied:
If I get it right, ptrauth_returns do not influence the ABI and ptrauth_calls do, thus it may be trivial to enable pac-ret in one's own build, but ptrauth_calls have to be consistently enabled for the whole runtime.

kovdan01 commented on libunwind/src/Registers.hpp:
```diff
@@ -1877,6 +1946,32 @@ inline Registers_arm64::Registers_arm64(const void *registers) {
   memcpy(_vectorHalfRegisters,
          static_cast<const uint8_t *>(registers) + sizeof(GPRs),
          sizeof(_vectorHalfRegisters));
+#if __has_feature(ptrauth_calls)
```
It's probably worth adding a comment explaining this piece of code

ojhunt replied:
- [ ] seems reasonable

kovdan01 commented on libunwind/src/Registers.hpp:
```diff
@@ -1845,10 +1871,53 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
+  }
+
+  // arm64_32 and i386 simulator hack
+  void loadAndAuthenticateLinkRegister(uint32_t srcLinkRegister,
```
I'm not sure that I get the point of this function overload here. Could you please clarify what do you mean by "arm64_32 and i386 simulator hack"? Deleting this locally did not harm compilation, and I'm not even sure that it's used somewhere.

ojhunt replied:
This is code we had downstream, I'm not sure of the reason for it, however if we switch to a model where we just gate the use of these functions on targeting arm64e and don't require every arch provide stub funcs I think that becomes moot?

ojhunt replied:
resolving this by switching to a single loadAndAuthenticateLinkRegister with each call being wrapped in an ifdef

kovdan01 commented on compiler-rt/lib/profile/InstrProfilingValue.c:
```diff
@@ -83,7 +83,13 @@ __llvm_profile_iterate_data(const __llvm_profile_data *Data) {
+#if __has_feature(ptrauth_calls)
+  // This is only used for tests where we compare against what happens to be
+  // signed pointers.
+  FP = ptrauth_sign_unauthenticated(FP, VALID_CODE_KEY, 0);
```
`VALID_CODE_KEY` is not defined, have you forgotten that? In some tests, there is `#define VALID_CODE_KEY 0` definition.

kovdan01 commented on compiler-rt/lib/profile/InstrProfilingValue.c:
```diff
@@ -83,7 +83,13 @@ __llvm_profile_iterate_data(const __llvm_profile_data *Data) {
+#if __has_feature(ptrauth_calls)
+  // This is only used for tests where we compare against what happens to be
+  // signed pointers.
+  FP = ptrauth_sign_unauthenticated(FP, VALID_CODE_KEY, 0);
```
In order to use `ptrauth_sign_unauthenticated`, you need to include ptrauth.h:

```
#if __has_include(<ptrauth.h>)
#include <ptrauth.h>
#endif
```

kovdan01 replied:
> @kovdan01 @asl do you folk have a vm setup you can test on? I realize I'm currently limited to building for darwin so it's easy for me to luck out on implicit/transitive includes from other places.

@ojhunt Yeah, we have that and we can run llvm-test-suite and some other private pauth-specific tests. And, of course, we have a build of pauth-enabled linux toolchain configured.

Basically, having these is the reason why I'm able to add such comments in this PR :)

ojhunt replied:
@kovdan01 I mean "do you have step by step instructions for a muppet like me that would make it possible for me to build/test locally" :D

kovdan01 replied:
> @kovdan01 I mean "do you have step by step instructions for a muppet like me that would make it possible for me to build/test locally" :D

@ojhunt Ah, thanks for clarifying your question :) Unfortunately, right now our build scripts, instructions, tests, ... are not available publicly. We'll definitely publish them, but right now I guess that we would be happy to test your changes on our side manually. This would, of course, slow down the work on the PR, but at this point it looks like that many of the comments we left actually are relevant for all targets, not only for linux. So, at least for such comments you'll be able to test things using your system w/o our linux-specific stuff.

kovdan01 commented on libunwind/src/libunwind.cpp:
```diff
@@ -126,6 +130,36 @@ _LIBUNWIND_HIDDEN int __unw_set_reg(unw_cursor_t *cursor, unw_regnum_t regNum,
+        if (ptrauth_auth_and_resign((void *)pc, ptrauth_key_return_address, sp,
+                                    ptrauth_key_return_address,
+                                    sp) != (void *)pc) {
+          _LIBUNWIND_LOG("Bad unwind through arm64e (0x%llX, 0x%llX)->0x%llX\n",
```
I believe that this check would be cross-platform and will do the trick for ELF platforms as well, so it's probably worth to re-phrase this and avoid arm64e term but say something in generic pauth terms.

ojhunt commented on libunwind/src/UnwindRegistersRestore.S:
```diff
@@ -694,7 +705,12 @@ DEFINE_LIBUNWIND_FUNCTION(__libunwind_Registers_arm64_jumpto)
 Lnogcs:
 #endif
+
+#if __has_feature(ptrauth_calls)
+  retab
+#else
```
see earlier comment - do we want to just adopt a single flag instead of the current mix of ptrauth_calls and ptrauth_returns ?

atrosinenko replied:
By the way, if switching to `ptrauth_returns` here, we most probably want to retain compatibility with Armv8.2-a by replacing
```
#if __has_feature(ptrauth_calls)
  retab
#else
  ret    x30                    // jump to pc
#endif
```
with
```
#if __has_feature(ptrauth_returns)
  autibsp
#endif
  ret    x30                    // jump to pc
```
That is, simply insert `autibsp` into the existing code.

kovdan01 commented:
@ojhunt JFYI: I've published a couple of small fixes I've applied locally at https://github.com/kovdan01/llvm-project/commits/pointer-authenticated-unwinding-fix0/. Maybe this would be useful

ojhunt commented on libcxxabi/src/cxa_personality.cpp:
```diff
@@ -541,7 +554,33 @@ struct scan_results
+// it impossible to directly cast them without breaking the authentication,
+// as a result we need this pair of helpers.
+template <typename PtrType>
+void set_landing_pad_as_ptr(scan_results& results, const PtrType& out) {
```
@kovdan01 ah yeah, unfortunately that's not sound, but for dumb reasons. The reason I've passed by reference is because the intent is for PtrType to be a ptrauth qualified type on arm64e, and we use a reference to ensure that it remains protected.

It might be possible to make this clearer once we have the schema queries upstreamed, because then we could make it explicit that this expects PtrType to be an address discriminated __ptrauth qualified pointer when compiling on arm64e (or whatever other platforms add ptrauth in future)

ojhunt commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -207,7 +211,8 @@ bool DwarfInstructions<A, R>::isReturnAddressSignedWithPC(A &addressSpace,
 template <typename A, typename R>
-int DwarfInstructions<A, R>::stepWithDwarf(A &addressSpace, pint_t pc,
+int DwarfInstructions<A, R>::stepWithDwarf(A &addressSpace,
+                                           typename R::link_reg_t &pc,
```
@kovdan01 as with the other case it's because link_reg_t is a ptrauth'd type and where intentionally keeping it protected.

ojhunt commented on libunwind/src/Registers.hpp:
```diff
@@ -1845,10 +1871,53 @@ class _LIBUNWIND_HIDDEN Registers_arm64 {
+    // with the SP
+    *referenceAuthedLinkRegister = (uint64_t)ptrauth_auth_data(
+        (void *)inplaceAuthedLinkRegister, ptrauth_key_return_address,
+        _registers.__sp);
```
I'll need to work out whether this is me screwing up the lifting of the PR, or me screwing up the authentication changes during cleanup, or something else entirely.

atrosinenko commented on compiler-rt/lib/builtins/gcc_personality_v0.c:
```diff
@@ -30,6 +30,46 @@ EXCEPTION_DISPOSITION _GCC_specific_handler(PEXCEPTION_RECORD, void *, PCONTEXT,
+#include <ptrauth.h>
+#endif
+
+#if defined(__APPLE__) && __has_feature(ptrauth_qualifier)
```
I am building this for `aarch64-linux-pauthtest` now and I wonder if it is beneficial to decrease the number of hardcoded `defined(__APPLE__)` guards? Something like
```cpp
#if defined(__APPLE__)
#define EXTRA_PTRAUTH_HARDENING 1
#else
#define EXTRA_PTRAUTH_HARDENING 0
#endif
```
near the top of the file and then use `EXTRA_PTRAUTH_HARDENING` throughout the code instead of `defined(__APPLE__)`, so that other platforms can opt-in to this hardening by adjusting a single line.

Though, the above `EXTRA_PTRAUTH_HARDENING` macro still have to be defined multiple times: separate definitions would probably be required for `compiler-rt/lib/builtins/gcc_personality_v0.c`, `libcxxabi` and `libunwind`.

ojhunt replied:
yeah, I wonder if there's any existing precedence for cmake variables the apply to all the runtimes?

Also, given some of the other issues you've hit with building don't expend too much effort trying to build/test - I need to work out what parts I've managed to screw up while building out this PR - some of the issues you've hit seem likely to be me screwing up during the preparation of the PR and I don't want you folk to slog through issues caused by my muppetry

kovdan01 commented on libcxxabi/include/__cxxabi_config.h:
```diff
@@ -109,4 +110,38 @@
+#  include <ptrauth.h>
+#endif
+
+#if defined(__APPLE__) && __has_feature(ptrauth_qualifier)
```
It looks like that `__has_feature(ptrauth_qualifier)` should be replaced with `__has_extension(ptrauth_qualifier)`. It is defined as `EXTENSION` in clang/include/clang/Basic/Features.def, and existing tests use `__has_extension` instead of `__has_feature` for this.

ojhunt commented:
I'll be back to working on this later today - there's a pile of feedback, and a bunch of nonsense that I managed to accidentally accrue and/or re-format when pushing the PR so it will take a little time to unbreak that. It may be easiest to force push a repaired PR first and then address comments.

I also realize that some of the breakage might be due to mismatched __APPLE__ vs has_feature(actual_feature)

atrosinenko reviewed (commented):
I have tested this patch on Linux with `aarch64-linux-pauthtest` triple and custom-built sysroot, here are the results.

In my setup, the tests pass on the following modified branch: https://github.com/atrosinenko/llvm-project/commits/pointer-authenticated-unwinding-fix1/. It is based on https://github.com/kovdan01/llvm-project/commits/pointer-authenticated-unwinding-fix0/ with the following changes:
* Currently, several parts of the patch are only enabled on Apple targets. For testing purposes, I unconditionally enabled parts of this patch guarded by `defined(__APPLE__)`.
* It turned out that `__has_feature(ptrauth_qualifier)` does not work in mainline, so I replaced it with `__has_feature(ptrauth_intrinsics)`. I expect the correct solution proposed by @kovdan01 (replacing with `__has_extension(ptrauth_qualifier)`) to behave identically.
* The change to `compiler-rt/lib/profile/InstrProfilingValue.c` seems to be unrelated to libunwind
* The most meaningful change is updating signing schemas is several places - see inline comments. Additionally, I had to update a few other places that I cannot add inline comments to
  - in `Registers_arm64::getRegister` and `Registers_arm64::setRegister`, replace direct references to `_registers.__pc` with calls to `getIP()` and `setIP()`, correspondingly
  - disable the existing pauth-related logic in `DwarfInstructions<A, R>::stepWithDwarf`

The fact I'm worried about is whether implicit signing and authentication on accesses to `__ptrauth`-qualified fields may introduce signing or authentication oracles usable by an attacker, since many values stored to these fields are initially non-signed. This is possibly mitigated by the fact that all these fields use address diversity with distinct integer discriminators and/or the original values are taken from read-only memory. On the other hand, discriminator computation, auth / sign intrinsic and load / store to memory are currently three separate operations when accessing a `__ptrauth`-qualified field, thus spilling of intermediate values to the stack is possible. Furthermore, even if the non-signed value originates from a read-only memory, this is not expressed in LLVM IR terms, thus the optimization pipeline may transform sensitive instruction sequences in an unsafe way.

ojhunt commented on libcxxabi/src/cxa_personality.cpp:
```diff
@@ -557,7 +596,19 @@ set_registers(_Unwind_Exception* unwind_exception, _Unwind_Context* context,
+  const auto existingDiscriminator = ptrauth_blend_discriminator(
+      &results.landingPad, ptrauth_string_discriminator(_LIBCXXABI_PTRAUTH_SCANRESULT_LANDINGPAD_DISC));
+  unw_word_t newIP =
+      (unw_word_t)ptrauth_auth_and_resign(*(void**)&results.landingPad, _LIBCXXABI_PTRAUTH_KEY, existingDiscriminator,
```
sigh - this code was using the pointer auth schema queries and I must have screwed up while converting it, however this can be mitigated somewhat by moving a bunch of these constants into ptrauth.h rather than duplicating each locally which is really what is causing this.

asl commented:
> The fact I'm worried about is whether implicit signing and authentication on accesses to `__ptrauth`-qualified fields may introduce signing or authentication oracles usable by an attacker, since many values stored to these fields are initially non-signed. This is possibly mitigated by the fact that all these fields use address diversity with distinct integer discriminators and/or the original values are taken from read-only memory. On the other hand, discriminator computation, auth / sign intrinsic and load / store to memory are currently three separate operations when accessing a `__ptrauth`-qualified field, thus spilling of intermediate values to the stack is possible. Furthermore, even if the non-signed value originates from a read-only memory, this is not expressed in LLVM IR terms, thus the optimization pipeline may transform sensitive instruction sequences in an unsafe way.

I think we will need to revise / reinvestigate this issue after this PR and see if we can move forward with more robust approach.

atrosinenko commented on libunwind/src/UnwindRegistersRestore.S:
```diff
@@ -681,7 +681,18 @@ DEFINE_LIBUNWIND_FUNCTION(__libunwind_Registers_arm64_jumpto)
+  ldr    x16,     [x0, #0x0F8]  // load sp into scratch
+  ldr    lr,      [x0, #0x100]  // restore pc into lr
+
+#if __has_feature(ptrauth_calls)
+  // The LR is signed with its address inside the register state.  Time
+  // to resign to be a regular ROP signed pointer
+  add    x1, x0, #0x100
+  autib  lr, x1
+  pacib  lr, x16  // signed the scratch register for sp
+#endif
```
Likewise, we should probably use `autib1716` and `pacib1716` here to retain compatibility with Armv8.2-a.

asl replied:
I do not think so. Overall pauth features being enabled imply proper ARM v8.3

kovdan01 replied:
Closing in favor of issue #160114

To be done as a follow-up

atrosinenko commented on libunwind/src/AddressSpace.hpp:
```diff
@@ -129,26 +129,38 @@ struct UnwindInfoSections {
 #endif
 #if defined(_LIBUNWIND_SUPPORT_COMPACT_UNWIND)
-  uintptr_t       compact_unwind_section;
-  size_t          compact_unwind_section_length;
+  uintptr_t __LIBUNWIND_PTRAUTH_RI_PDD(
+      "UnwindInfoSections::compact_unwind_section") compact_unwind_section = 0;
+  size_t __LIBUNWIND_PTRAUTH_RI_PDD(
+      "UnwindInfoSections::compact_unwind_section_length")
+      compact_unwind_section_length = 0;
 #endif
 #if defined(_LIBUNWIND_ARM_EHABI)
-  uintptr_t       arm_section;
-  size_t          arm_section_length;
+  uintptr_t __LIBUNWIND_PTRAUTH_RI_PDD("UnwindInfoSections::arm_section")
+      arm_section = 0;
+  size_t __LIBUNWIND_PTRAUTH_RI_PDD("UnwindInfoSections::arm_section_length")
+      arm_section_length = 0;
```
These two fields doesn't seem to be used on AArch64.

ojhunt commented:
Re: the oracle concerns - there are some protections we have, but in general I think the current pointer auth semantics/diagnostics make it too easy to silently introduce oracles. Additionally a lot of the current unwind pointerauth work is focused on protecting the ABI boundaries in the knowledge we can improve the internals over time.

I've once again acquired some higher priority work but I should be getting back to this tonight or over the weekend.

ojhunt commented:
> @ojhunt Just in case: have you had a chance to look into comments left above? I would be glad to provide any help or assistance if it's needed. It would be very nice if we could get this fixed and merged before upcoming release.

I have a significant set of changes I'm working through, hopefully addressing your comments (moving a lot of the constants into ptrauth.h makes a lot of sense but requires a pile of testing and world builds to ensure consistency - I've repeatedly typo'd things so am being very paranoid), at the same time I've been side tracked by the objc work.

ojhunt commented:
Force push+rebase because I've significantly reworked and tidied the changes.

I've centralized the schema definitions in each library, and made them explicit rather than constructed at each site.

I've also replaced the __APPLE__ guards with feature guards which seems much more reasonable - the places I thought __APPLE__ made sense we really intra library PAC which makes that kind of silly.

ojhunt commented:
(if linux folk could retest to see how badly I have broken everything that would be great)

kovdan01 commented:
> (if linux folk could retest to see how badly I have broken everything that would be great)

@ojhunt A couple of quick observations:

1. It looks like that you forgot to add explicit declarations of copy constructor and assignment operator for `Registers_arm64`. Now, compiler complains about explicit definitions for these while they are implicitly defined.

2. Why do we need `static_assert(!__has_extension(ptrauth_qualifier));` under ifdefs for `_LIBUNWIND_USE_DL_ITERATE_PHDR`, `_LIBUNWIND_SUPPORT_DWARF_INDEX` and `_LIBUNWIND_ARM_EHABI`? These are failing for me (particularly the first two ones), and I'm not sure I understand why these static assertions are needed.

3. The PR title contains a reference to an old PR #138571, it's probably not needed

ojhunt commented:
> > (if linux folk could retest to see how badly I have broken everything that would be great)
> 
> @ojhunt A couple of quick observations:
> 
> 1. It looks like that you forgot to add explicit declarations of copy constructor and assignment operator for `Registers_arm64`. Now, compiler complains about explicit definitions for these while they are implicitly defined.

Hmmm, what is the configuration you're building with? it's possible a mismerge, but also possible I just screwed up builds.

> 2. Why do we need `static_assert(!__has_extension(ptrauth_qualifier));` under ifdefs for `_LIBUNWIND_USE_DL_ITERATE_PHDR`, `_LIBUNWIND_SUPPORT_DWARF_INDEX` and `_LIBUNWIND_ARM_EHABI`? These are failing for me (particularly the first two ones), and I'm not sure I understand why these static assertions are needed.

I was verifying code I was building was being built as I expected, I'll remove those assertions - I realized while the original PR applied the `__ptrauth` qualifier to fields we don't have a testable configuration for, and I scattered a bunch of static asserts to verify what all our configurations were hitting because if they weren't I could not test them :-/ 

> 3. The PR title contains a reference to an old PR [[runtimes?][PAC] harden unwinding #138571](https://github.com/llvm/llvm-project/issues/138571), it's probably not needed

Righto

ojhunt commented:
@kovdan01 ah I just see the bots all hate me so I think I've mismerged :D

kovdan01 commented on libcxxabi/src/cxa_personality.cpp:
```diff
@@ -21,6 +21,44 @@
+#define __ptrauth_scan_results_landingpad \
+   __ptrauth(__ptrauth_scan_results_landingpad_key, 1, __ptrauth_scan_results_landingpad_disc)
+#define __ptrauth_scan_results_landingpad_intptr \
+   __ptrauth_restricted_intptr(__ptrauth_scan_results_landingpad_key, 1, \
```
I think that this should go under `#if __has_extension(ptrauth_restricted_intptr_qualifier)` (just like in other places touched by this PR, see libunwind.h for example). As far as I can see, `__ptrauth_restricted_intptr` is not yet supported by mainline, and using that causes compile errors on my side.

ojhunt replied:
it actually needs the ptrauth/ptrauth_intptr macros to deal with that separate spelling. Will add it to my current refactoring fixes.

kovdan01 commented on libcxxabi/include/__cxxabi_config.h:
```diff
@@ -103,10 +104,68 @@
 #define _LIBCXXABI_DTOR_FUNC
 #endif
 
-#if __cplusplus < 201103L
```
I suppose this was removed accidentally during merging. W/o this, the build is just broken.

ojhunt replied:
by failing to compile we ensure there are no memory safety bugs. :D

atrosinenko commented on libcxxabi/include/__cxxabi_config.h:
```diff
@@ -103,10 +104,68 @@
+
+#  endif
+
+// Helper wrappers for pointer auth qualifiers because we use a lot of variants
+// Suffixes:
+//  * _RI : qualifier is __ptrauth_restricted_intptr
+//  * PDD : key is ptrauth_key_process_dependent_data
+//  * FN  : key is ptrauth_key_function_pointer
+#define _LIBCXXABI_PTRAUTH_PDD(__discriminator)                                                                        \
+  _LIBCXXABI_PTRAUTH(ptrauth_key_process_dependent_data, /*__address_discriminated=*/1, __discriminator)
+#define _LIBCXXABI_PTRAUTH_FN(__discriminator)                                                                         \
+  _LIBCXXABI_PTRAUTH(ptrauth_key_function_pointer, /*__address_discriminated=*/1, __discriminator)
+#define _LIBCXXABI_PTRAUTH_RI_PDD(__discriminator)                                                                     \
+  _LIBCXXABI_PTRAUTH_RESTRICTED_INTPTR(ptrauth_key_process_dependent_data, /*__address_discriminated=*/1,              \
+                                       __discriminator)
+
```
This seems to be unused now.

ojhunt replied:
yup I saw that as well while going through the "what did I mess up in this latest merge", doing a bunch of testing at the moment

ojhunt commented on compiler-rt/lib/builtins/gcc_personality_v0.c:
```diff
@@ -30,6 +30,50 @@ EXCEPTION_DISPOSITION _GCC_specific_handler(PEXCEPTION_RECORD, void *, PCONTEXT,
+// of maintaining a record of how they were originally produced.
+
+// ptrauth_string_discriminator("__gcc_personality_v0'funcStart") == 0xDFEB
+#define __ptrauth_gcc_personality_func_start \
```
Yeah the problem here is that I was finding the length of the qualifiers was actively making it hard to read the code.

I need to go through and come up with some more compact but still understandable.

atrosinenko commented:
By the way, for those reviewers who wonder why `__ptrauth_restricted_intptr` qualifier is sometimes used instead of `__ptrauth`, here is the discussion: #137580.

ojhunt committed:
[runtimes][PAC] Harden unwinding when possible (#138571)

This hardens the unwinding logic and datastructures on systems
that support pointer authentication.
The approach taken to hardening is to harden the schemas of as many
high value fields in the myriad structs as possible, and then also
explicitly qualify local variables referencing privileged or security
critical values.
This does introduce ABI linkage between libcxx, libcxxabi, and
libunwind but those are in principle separate from the OS itself
so we've kept the schema definitions in the library specific headers
rather than ptrauth.h

ojhunt commented on libcxxabi/src/cxa_personality.cpp:
```diff
@@ -21,6 +21,44 @@
+#endif
+
+
+#include "libunwind.h"
```
have removed this now

ojhunt commented on libunwind/src/UnwindLevel1.c:
```diff
@@ -597,6 +611,18 @@ _LIBUNWIND_EXPORT uintptr_t _Unwind_GetIP(struct _Unwind_Context *context) {
   unw_word_t result;
   __unw_get_reg(cursor, UNW_REG_IP, &result);
+
+#if __has_feature(ptrauth_calls)
+  // If we are in an arm64e frame, then the PC should have been signed with the
+  // sp
+  {
+    unw_word_t sp;
+    __unw_get_reg(cursor, UNW_REG_SP, &sp);
+    result = (unw_word_t)ptrauth_auth_data((void *)result,
+                                           ptrauth_key_return_address, sp);
+  }
+#endif
```
I think we'll need to go through the pr to determine which of the ptrauth_calls should be ptrauth_returns as our codebase assumes they're always the same

ojhunt committed:
Updating to include some fixes I lost in the last update

At some point I lost the changes to loadAndAuthenticateLinkRegister
I also updated schema names in libunwind to be more consistent
Finally while looking at the total diff I saw some places that
the formatting could be improved.

atrosinenko reviewed (commented):
@ojhunt Thank you for the update! I like the idea of not checking for `__has_include(<ptrauth.h>)` separately, as it was done in your latest version of `gcc_personality_v0.c`. I wonder if other instances of `#if __has_include(<ptrauth.h>)` are worth replacing with unconditionally including `ptrauth.h` as soon as it is decided to use its features according to `__has_feature(...)` tests (and erroring out in case this file is absent). Are there any expected use cases when absence of `ptrauth.h` has to be handled gracefully?

ojhunt commented on libunwind/src/CompactUnwinder.hpp:
```diff
@@ -680,11 +682,19 @@ int CompactUnwinder_arm64<A>::stepWithCompactEncodingFrame(
+  // arm64e it's the discriminator used for those registers.
   registers.setSP(fp + 16);
+
+  Registers_arm64::reg_t oldfp = addressSpace.get64(fp);
+
+  // fp points to old fp
+  registers.setFP(oldfp);
```
So this is an area I'm unsure of - I thought the addressSpace abstraction was for cross process. ... something? But I don't know if that was just a hypothetical future concept or is actually used somewhere?

It's also possible that this is just left over from a bunch of dead code I excised while preparing this

ojhunt replied:
There was a period we experimented with frame signing (it was gated on a ptrauth_frame feature test), and while prepping these I removed the defunct code to avoid the appearance of it existing, or providing a no longer verified to be working feature, it's possible some of the weird ordering isn't needed for this anymore.

I'll update the comments to remove the reference to needing FP stuff.

ojhunt replied:
Cleaned this up

ojhunt committed:
Build fix due to not testing the build after apply review feedback \o/

ojhunt commented on libcxxabi/include/__cxxabi_config.h:
```diff
@@ -103,6 +103,52 @@
 #define _LIBCXXABI_DTOR_FUNC
 #endif
 
+#if __has_include(<ptrauth.h>)
+#  include <ptrauth.h>
+#endif
```
So this is something I'm unsure about - there's a lot of checks for the header scattered everywhere but also plenty of places that rely on the feature existing and blindly including the header.

I'm ok putting the ptrauth.h includes unconditionally inside the feature check on the basis that if the header isn't present the ptrauth enabled paths will fail to build

ojhunt commented:
Ok a bunch of these errors are because clang is misdiagnosing this scenario

```cpp
namespace {
template <class T> void foo(T*){}
}

template <class T> void bar(T* t) {
    foo(t);
}
```
as not using the template function `foo`

ojhunt committed:
work around a bug where clang reports a template function as unused

ojhunt committed:
And another one

ojhunt committed:
Add all the required reg_t and link_reg_t decls, and unify location

ojhunt committed:
This erroneous error is annoying

ojhunt committed:
Remove header include that is unnecessary but apparently works in some places

ojhunt committed:
Update for feature check changes

kovdan01 commented:
Thanks @ojhunt for updates! These resolve some of the previously found issues, but many problems are still present (both previously reported and newly found). I tried to resolve all the threads which are now no longer relevant, so please treat all the opened threads as issues which need your attention.

I would be happy to provide any help - just let me know if it's needed.

Here are the issues I'd want to highlight - these should probably be resolved first.

1. Build issues due to missing `unw_word_t` and `UNW_REG_SP` declarations in libcxxabi/src/cxa_personality.cpp: https://github.com/llvm/llvm-project/pull/143230#discussion_r2316986679

2. Fix personality pointer authentication on non-Apple targets: https://github.com/llvm/llvm-project/pull/143230#discussion_r2318636296

3. Fix handling signed RA in `DwarfInstructions<A, R>::stepWithDwarf`: https://github.com/llvm/llvm-project/pull/143230#discussion_r2318709982

4. Look through ptrauth_calls/ptrauth_returns/other ptrauth checks and ensure that they are correct. On arm64e, there is no difference, while correct checks might be important for other ptrauth-enabled ABIs when having feature X does not necessarily imply having feature Y

ojhunt commented on libcxxabi/src/cxa_personality.cpp:
```diff
@@ -557,7 +630,21 @@ set_registers(_Unwind_Exception* unwind_exception, _Unwind_Context* context,
   _Unwind_SetGR(context, __builtin_eh_return_data_regno(1),
                 static_cast<uintptr_t>(results.ttypeIndex));
+#if __has_feature(ptrauth_qualifier)
+  auto stackPointer = _Unwind_GetGR(context, UNW_REG_SP);
```
Sigh, i've been fighting some local build issues and was wanting to get bot feedback while getting my local build working again (see https://github.com/llvm/llvm-project/pull/156607 that I ran into because I was accidentally building clang fat). I should have commented that this was not ready for re-review (though the feedback is needed, it would hopefully mitigate manually finding failures)

Including libunwind.h is necessary -- pointer auth makes these ABI linked, and while we _could_ redeclare these in cxxabi that would obscure the linkage _and_ mean changes are  not automatic and silently break.

ojhunt commented on libcxxabi/src/cxa_personality.cpp:
```diff
@@ -541,7 +588,33 @@ struct scan_results
+// it impossible to directly cast them without breaking the authentication,
+// as a result we need this pair of helpers.
+template <typename PtrType>
+[[maybe_unused]] void set_landing_pad_as_ptr(scan_results& results, const PtrType& out) {
```
you did not get a warning when not compiling arm64e support? It's possible trunk clang has fixed this?

ojhunt commented on libunwind/src/UnwindCursor.hpp:
```diff
@@ -1047,18 +1051,26 @@ class UnwindCursor : public AbstractUnwindCursor{
+                               const UnwindInfoSections &sects,
+                               uint32_t fdeSectionOffsetHint = 0);
   int stepWithDwarfFDE(bool stage2) {
+#if __has_extension(ptrauth_calls)
```
sigh, I thought I had found all of the has_extension and has_feature checks everywhere (except for those places that define the macros we use -- the few places it seemed unnecessary should have been updated with `__PTRAUTH__` checks.

ojhunt commented on compiler-rt/lib/builtins/crtbegin.c:
```diff
@@ -16,6 +16,12 @@
 #include <ptrauth.h>
 #endif
 
+#if __has_feature(ptrauth_calls) || defined(__PTRAUTH__)
```
I found this while updating the __PTRAUTH__ guards -- I would be ok leaving this as a task for you folk as I think this is a linux specific file?

ojhunt commented:
hmmm, I thought `ptrauth_calls` was one of the checks guarded by a darwin target so I might revert the `|| defined(__PTRAUTH__)` changes

ojhunt committed:
yet more cleaning up

ojhunt committed:
cleanup compiler-rt formatting

ojhunt committed:
update libcxxabi formatting

ojhunt commented:
Updating cxxabi and compiler-rt formatting as they seem to be set up such that clang-format matches the existing style

ojhunt commented:
> > > * Fix personality pointer authentication on non-Apple targets: [[runtimes][PAC] Harden unwinding when possible #143230 (comment)](https://github.com/llvm/llvm-project/pull/143230#discussion_r2318636296)
> > 
> > 
> > I'm not sure what's going on in this diff as it does not appear related to changes in this PR? Is it just a general issue that already exists?
> 
> @ojhunt Well, I'd rather say that it's not an issue that already exists, it's just missing functionality - libunwind does not support ptrauth-hardened personality pointer, and this is true for both signed personality pointer on Apple and Linux. We just have different signing schemes, and the signing scheme on Linux is defined in this PR: #119361
> 
> If you prefer not to include support for signed personality on Linux, I would be happy to submit support for that as a follow-up PR as soon as this one is merged (since logic for Linux is similar to logic for Apple, we just have different discriminators). But, if we take this approach, let's at least add a TODO note mentioning that signed personality now only works for Apple's arm64e, and support for Linux will be implemented in future.
> 
> Probably, we can use smth like this:
> 
> ```
> #if __has_feature(ptrauth_calls) // as discussed in discord, let's replace that with some macro for calls+returns
>         if (personality) {
> #if __APPLE__
>           const auto discriminator = ptrauth_blend_discriminator(
>               &cieInfo->personality, __ptrauth_unwind_cie_info_personality_disc);
>           void *signedPtr = ptrauth_auth_and_resign(
>               (void *)personality, ptrauth_key_function_pointer, resultAddr,
>               ptrauth_key_function_pointer, discriminator);
>           personality = (pint_t)signedPtr;
> #else
> #error "Signed personality function pointer is not supported for non-Apple targets"
> // TODO: Implement support for Linux. The signing scheme of personality pointer
> // differ for Linux and Apple targets. For Linux targets, the signing scheme is
> // defined in https://github.com/llvm/llvm-project/pull/119361
> #endif
>         }
> #endif
> ```
> 
> Which approach do you prefer? Is it applying the suggestion from commit [e2f8b9d](https://github.com/llvm/llvm-project/commit/e2f8b9d3859eff96442ce04662aefb40debbef3f) or postponing the Linux support and using the code snippet like above?
> 
> I'm OK with both, just please let me know which is more comfortable for you.

I _think_ doing it as a follow up makes sense: it's not a matter of me not wanting to do this for linux, it's that I cannot test it at all. Doing it as a follow up would also give you the constants block to put the definition rather than having it just slotted in the code. I've found moving to the constant blocks to be incredibly useful for being able to see where/how they cross ABI boundaries.

ojhunt commented:
> > I _think_ doing it as a follow up makes sense: it's not a matter of me not wanting to do this for linux, it's that I cannot test it at all. Doing it as a follow up would also give you the constants block to put the definition rather than having it just slotted in the code. I've found moving to the constant blocks to be incredibly useful for being able to see where/how they cross ABI boundaries.
> 
> @ojhunt Thanks! So could you please adopt the code snippet from my previous comment to make things explicitly unsupported for Linux?
> 
> And also I kindly ask you to fix p.4 from [#143230 (comment)](https://github.com/llvm/llvm-project/pull/143230#issuecomment-3249013139) (we've already discussed that in Discord). Please let me know if you need any help.
> 
> A couple of things are on my side:
> 
> 1. Compile a local pac-ret-enabled toolchain. This would allow you to fix p.3 from the same comment. I'll let you know when it's ready.

In _principle_ I could set up this configuration right? (Obviously not actually test it, but in principle at least be able to verify it builds before pushing the update)

> 2. Look through other comments in this PR (they have lower priority since they are not related to any runtime/compile-time failures). I'll highlight the ones which I think are worth fixing in scope of this PR (if any).
> 
> Please let me know if I've missed smth or if you have any objections on this plan.

GH makes finding these things so utterly miserable. I hate it so very much.

atrosinenko reviewed (commented):
It would be probably useful to explicitly list the configurations that we would like to support at the end of the day. Some combinations are probably unsupported or not planned for the first version of the patch. I have already left similar comments before, but let me summarize them in a more structured fashion. Corrections are welcome :)

There seems to be three different aspects independent from each other. The first aspect is pac-ret hardening mode the calling code was built with. If my understanding is correct, pac-ret is largely considered as not affecting the ABI (because signing and authentication code belongs to the same function) and thus the mode can be arbitrarily chosen per shared object, per-library or technically even per-function. Thus, this mode should probably be inferred not from the preprocessor macroses observed at libunwind build time, but from the DWARF information (which raises the question of when we can or cannot trust this information under Pointer Authentication threat model). See [`aadwarf64` document](https://github.com/ARM-software/abi-aa/releases/tag/2025Q1) for the description of CFA operations controlling `RA_SIGN_STATE` pseudo register. Furthermore, the *key* being used by the particular stack frame is communicated via optional character `B` in the CIE augmentation string (described in the same document). There already exist significant amount of code in [`DwarfInstructions<A, R>::stepWithDwarf`](https://github.com/llvm/llvm-project/blob/47c1b650626043f0a8f8e32851617201751f9439/libunwind/src/DwarfInstructions.hpp#L210) - I wonder if it could be reused in a secure manner.

The second aspect is ABI-affecting variations of unwind information format. One thing I do know about is signing of the pointer to the personality function, but I don't have the complete list of structs and fields in my mind. It would probably be useful to explicitly enumerate the minimal set of "public" fields whose signing schemas we should choose in accordance with the ABI. After all, these details should probably end up being documented in some specification sooner or later.

The third aspect is which internal fields are signed and when. This is more a matter of the targeted ISA version and best-effort guess of the desired "protection mode" of the affected libraries. According to https://github.com/llvm/llvm-project/pull/143230#discussion_r2173023794, this could probably be as simple as "if `__has_feature(ptrauth_calls)` is true, then assume we target at least Armv8.3-a and enable signing of all the affected internal fields". (I don't remember, but maybe some rudimentary v8.2-compatible hardening of PC exists in DwarfInstructions as well.)

By the way, the separation of purely internal fields should also help to simplify the code: while ABI-defined signing schemas have to be described explicitly, for the internal struct fields and local variables it should be possible to define macroses like

```cpp
#define PTRAUTH_INTPTR(type, name) \
    type __ptrauth_gcc_personality_intptr(__ptrauth_gcc_personality_func_key,  \
    1, ptrauth_string_discriminator("__gcc_personality_v0'" #name)) name
```

or maybe embed the type into the macro name like `PTRAUTH_UINTPTR_T(name)`. Then it can be used along these lines:

```cpp
    PTRAUTH_INTPTR(uintptr_t, start) = readEncodedPointer(&p, callSiteEncoding);
```

This would additionally clarify the distinction between ABI-defined and arbitrarily chosen schemas. I wonder if this approach has its own important downsides that I don't see.

ojhunt commented on compiler-rt/lib/builtins/gcc_personality_v0.c:
```diff
@@ -30,6 +30,45 @@ EXCEPTION_DISPOSITION _GCC_specific_handler(PEXCEPTION_RECORD, void *, PCONTEXT,
 
+#if __has_feature(ptrauth_qualifier)
+#include <ptrauth.h>
+#if __has_feature(ptrauth_restricted_intptr_qualifier)
```
> 1. (To be fixed in scope of this PR) Comment in this PR about `ptrauth_restricted_intptr_qualifier`: TODO

No, that renders the library unbuildable in shipping apple clang.

kovdan01 commented on libcxxabi/src/cxa_exception.h:
```diff
@@ -47,10 +47,11 @@ struct _LIBCXXABI_HIDDEN __cxa_exception {
-    std::unexpected_handler unexpectedHandler;
-    std::terminate_handler  terminateHandler;
+    std::unexpected_handler _LIBCXXABI_PTRAUTH_FN("__cxa_exception::unexpectedHandler") unexpectedHandler;
+    std::terminate_handler _LIBCXXABI_PTRAUTH_FN("__cxa_exception::terminateHandler") terminateHandler;
```
Closing in favor of issue #160119

To be done as a follow-up

kovdan01 commented on compiler-rt/lib/builtins/gcc_personality_v0.c:
```diff
@@ -30,6 +30,45 @@ EXCEPTION_DISPOSITION _GCC_specific_handler(PEXCEPTION_RECORD, void *, PCONTEXT,
+#if __has_feature(ptrauth_qualifier)
+#include <ptrauth.h>
+#if __has_feature(ptrauth_restricted_intptr_qualifier)
+#define __ptrauth_gcc_personality_intptr(key, addressDiscriminated,            \
```
Closing in favor of https://github.com/llvm/llvm-project/pull/143230/files#r2369029642

To be done in scope of this PR

ojhunt committed:
Just make this a release mode failure

ojhunt commented on libunwind/src/libunwind.cpp:
```diff
@@ -118,22 +119,51 @@ _LIBUNWIND_HIDDEN int __unw_set_reg(unw_cursor_t *cursor, unw_regnum_t regNum,
+      // It is only valid to set the IP within the current function.
+      // This is important for ptrauth, otherwise the IP cannot be correctly
+      // signed.
+      [[maybe_unused]]unw_word_t stripped_value =
```
I may move the helpers - currently I've put a new ifdef, while I work out if there was a real reason for where I placed the helpers.

ojhunt committed:
Address comments

ojhunt committed:
Sigh, actually make sure it builds, also address warnings

kovdan01 commented on libcxxabi/src/cxa_exception.h:
```diff
@@ -47,10 +47,10 @@ struct _LIBCXXABI_HIDDEN __cxa_exception {
     void *(_LIBCXXABI_DTOR_FUNC *exceptionDestructor)(void *);
 #else
-    void (_LIBCXXABI_DTOR_FUNC *exceptionDestructor)(void *);
+    void(_LIBCXXABI_DTOR_FUNC *__ptrauth_cxxabi_exception_destructor exceptionDestructor)(void*);
```
Nit: revert unintended formatting change
```suggestion
    void (_LIBCXXABI_DTOR_FUNC *__ptrauth_cxxabi_exception_destructor exceptionDestructor)(void *);
```

ojhunt replied:
sigh I thought clang-format + checking/adjustment had actually got these ones right

kovdan01 commented on libcxxabi/src/cxa_exception.h:
```diff
@@ -79,16 +79,18 @@ struct _LIBCXXABI_HIDDEN __cxa_exception {
-    void (_LIBCXXABI_DTOR_FUNC *exceptionDestructor)(void *);
-    std::unexpected_handler unexpectedHandler;
-    std::terminate_handler terminateHandler;
+    void(_LIBCXXABI_DTOR_FUNC *__ptrauth_cxxabi_exception_destructor exceptionDestructor)(void*);
```
Nit: revert unintended formatting change
```suggestion
    void (_LIBCXXABI_DTOR_FUNC *__ptrauth_cxxabi_exception_destructor exceptionDestructor)(void *);
```

kovdan01 commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -300,12 +302,12 @@ int DwarfInstructions<A, R>::stepWithDwarf(A &addressSpace, pint_t pc,
-      // return address needs to be authenticated before the return address is
-      // restored. autia1716 is used instead of autia as autia1716 assembles
-      // to a NOP on pre-v8.3a architectures.
+#if defined(__ARM64E__)
```
I've not done local testing with the latest version of the PR, but it looks like that this changes implements *exactly* the opposite behavior of what we are trying to achieve :)

The code under this conditional serves for handling LR signed with pac-ret. And we actually need this to be guarded with smth like `#ifndef LIBUNWIND_PTRAUTH_CALLS_AND_RETURNS` to avoid executing this code when using Apple's arm64e/Linux's pauthtest (because this way, LR is signed as part of corresponding ABI and not by pac-ret).

If we put this under `#if defined(__ARM64E__)`, we entirely disable pac-ret handling for non-arm64e. This would break things which were working as intended before this change.

See my comment https://github.com/llvm/llvm-project/pull/143230#discussion_r2369419226 and commit with proposed fix ced8b99373c9b0756f1718999996f44a74bdf46d

Please let me know if any help is needed with fixing the issue.

kovdan01 commented:
@ojhunt Thanks for an update! I've resolved several threads, but some things should still be fixed.

Please let me know if you have any comments/questions or any help is needed from my side.

### Trivial style-related issues

These require almost zero effort to apply. Please consider fixing these first so we do not need to worry about them anymore.

1. https://github.com/llvm/llvm-project/pull/143230#discussion_r2388284017
2. https://github.com/llvm/llvm-project/pull/143230#discussion_r2388324262
3. https://github.com/llvm/llvm-project/pull/143230#discussion_r2388325718

### Non-trivial styling-related issues

Refactor preprocessor checks against ptrauth. See thread https://github.com/llvm/llvm-project/pull/143230#discussion_r2369029642 and proposed fix 644405b56cfa59dd3787119182df087dff6e756c

### Breaking change to be reverted

> > ### An issue causing a runtime crash
> > Do not mix pac-ret and ptrauth_returns. See thread https://github.com/llvm/llvm-project/pull/143230/files#r2369419226 and commit [ced8b99](https://github.com/llvm/llvm-project/commit/ced8b99373c9b0756f1718999996f44a74bdf46d).
> 
> Fixed with an arm64e guard.

This does not seem to fix the originally described issue, moreover, it looks like this breaks previously working behavior. See https://github.com/llvm/llvm-project/pull/143230#discussion_r2388356981.

ojhunt committed:
Updating for feedback

ojhunt commented on libunwind/src/DwarfParser.hpp:
```diff
@@ -313,6 +314,18 @@ bool CFI_Parser<A>::findFDE(A &addressSpace, pint_t pc, pint_t ehSectionStart,
+                                               T signed_personality) {
+  static_assert(sizeof(info->personality) == sizeof(signed_personality),
+                "Signed personality is the wrong size");
+  memmove((void *)&info->personality, (void *)&signed_personality,
```
Hmmm, other platforms haven't simply adopted the "handle overlap correctly everywhere" approach? :-/

kovdan01 replied:
> Hmmm, other platforms haven't simply adopted the "handle overlap correctly everywhere" approach? :-/

@ojhunt I honestly can't say for all the platforms, but I suppose that `memcpy` expresses the intention better than `memmove` here

ojhunt committed:
whoops, wrong test

ojhunt commented:
> @ojhunt Thanks for an update! I've resolved several threads, but some things should still be fixed.
> 
> Please let me know if you have any comments/questions or any help is needed from my side.
> 
> ### Trivial style-related issues
> These require almost zero effort to apply. Please consider fixing these first so we do not need to worry about them anymore.
> 
> 1. [[runtimes][PAC] Harden unwinding when possible #143230 (comment)](https://github.com/llvm/llvm-project/pull/143230#discussion_r2388284017)
> 2. [[runtimes][PAC] Harden unwinding when possible #143230 (comment)](https://github.com/llvm/llvm-project/pull/143230#discussion_r2388324262)
> 3. [[runtimes][PAC] Harden unwinding when possible #143230 (comment)](https://github.com/llvm/llvm-project/pull/143230#discussion_r2388325718)

I think fixed?

> 
> ### Non-trivial styling-related issues
> Refactor preprocessor checks against ptrauth. See thread [#143230 (comment)](https://github.com/llvm/llvm-project/pull/143230#discussion_r2369029642) and proposed fix [644405b](https://github.com/llvm/llvm-project/commit/644405b56cfa59dd3787119182df087dff6e756c)

That sounds ok

> 
> ### Breaking change to be reverted
> > > ### An issue causing a runtime crash
> > > Do not mix pac-ret and ptrauth_returns. See thread https://github.com/llvm/llvm-project/pull/143230/files#r2369419226 and commit [ced8b99](https://github.com/llvm/llvm-project/commit/ced8b99373c9b0756f1718999996f44a74bdf46d).
> > 
> > 
> > Fixed with an arm64e guard.
> 
> This does not seem to fix the originally described issue, moreover, it looks like this breaks previously working behavior. See [#143230 (comment)](https://github.com/llvm/llvm-project/pull/143230#discussion_r2388356981).

Minor detail :D

I think fixed correctly now (based on what you posted) but I need to verify the correct thing happens on arm64e (this will take a few days).

There's still the ifdef'd definition of the helper function in the current version, but I'll move that around in a bit.

ojhunt committed:
You would not believe how long it took to find this

ojhunt commented on libunwind/src/UnwindLevel1.c:
```diff
@@ -90,6 +90,18 @@
+    _Unwind_Personality_Fn __ptrauth_unwind_upi_handler *handler;
+  } u;
+  u.opaque_handler = (void *)&frameInfo->handler;
+  return *u.handler;
```
memcpy cannot be used -> these are address discriminated values, the entire reason for these hoops is to deal with this problem.

ojhunt replied:
The alternative is to force a copy + re-sign

ojhunt replied:
Actually, I think the issues that these workarounds were actually needed for have since been resolved.

ojhunt committed:
Addressing some feedback

* Add note for `__ptrauth_restricted_intptr`
* Rename confusing parameter in set_landing_pad_as_ptr
* Rename __ptrauth_unwind_pacret_personality_disc to
  __ptrauth_unwind_pauthtest_personality_disc

ojhunt committed:
cleaning up the [ab]use of unions

ojhunt committed:
Addressing myriad feedback

ojhunt committed:
Fixing remaining issues

Stop using unions to deal with the pointer auth cast semantics,
instead perform manual re-signing. This means the templated helper
functions no longer need to be templates and so don't need to be
in a separated non-extern "C" block.

Fixing up the comments.

ldionne reviewed (commented):
The patch looks fine to me but you should finish addressing comments by @kovdan01 .

I wonder, however, what kind of test coverage we have for your changes. Do we have any? In the current CI, are we even exercising these changes?

kovdan01 reviewed (commented):
@ojhunt Thanks for addressing the issues!

I've run tests on our side and I only see 2 issues at this point, and both of them seem to be fixed trivially. See my branch https://github.com/kovdan01/llvm-project/commits/pointer-authenticated-unwinding-fix-2025-10-14/

1. You use `ptrauth_function_pointer_type_discriminator` in UnwindLevel1.c, but it looks like this is not defined anywhere, which causes compile error. Please add a definition or just use `ptrauth_type_discriminator`. For my local tests, I've switched to `ptrauth_type_discriminator` to avoid compiler error - see commit https://github.com/kovdan01/llvm-project/commit/7a43f48bdac2bba68fc9f392361b44d586ca06d9

2. In `get_handler_function`, you assume that function pointer is signed with type discrimination enabled. This is not necessarily true, for example, this is not included in pauthtest ABI. Please add a corresponding check against `#if __has_feature(ptrauth_function_pointer_type_discrimination)` - see commit https://github.com/kovdan01/llvm-project/commit/57c46737ba26b28010443b66ea24e1a3dc6e9890

When these issues are resolved, this PR LGTM.

ojhunt commented:
> The patch looks fine to me but you should finish addressing comments by @kovdan01 .
> 
> I wonder, however, what kind of test coverage we have for your changes. Do we have any? In the current CI, are we even exercising these changes?

These changes do not effect behavior if there is no pointer authentication - if there is pointer authentication then these have to be consistent with the host platform so I'm not sure how to test conformance without adding what are in essence "expected fail" tests where we construct intentionally incorrect frames and attempt to unwind through them. Having this upstreamed would permitted arm64e apple silicon test bots to run which might be useful?

ojhunt committed:
formatting corrections

ojhunt committed:
A somewhat gross 'lets see if this is the bot issue' change

ojhunt commented:
gnahhhhhh why is [a9d9113](https://github.com/llvm/llvm-project/pull/143230/commits/a9d911322fe3020fd098ff0d6cbf6745b66b93e6) apparently fixing the armv8 build? That implies something it assuming Arm64Registers is a POD type, and I can't work out where :-/

ojhunt committed:
Ooooh, this might be it -- also it remains absurd that this isn't an error

ojhunt committed:
Let's stop doing this in the fragile field by field way

ojhunt committed:
Fix the build because I am clever

kovdan01 commented on libunwind/src/DwarfInstructions.hpp:
```diff
@@ -22,7 +22,6 @@
 #include "dwarf2.h"
 #include "libunwind_ext.h"
 
-
```
Nit: unintended formatting change

kovdan01 commented on libunwind/src/libunwind.cpp:
```diff
@@ -118,22 +118,62 @@ _LIBUNWIND_HIDDEN int __unw_set_reg(unw_cursor_t *cursor, unw_regnum_t regNum,
+        // This is important for ptrauth, otherwise the IP cannot be correctly
+        // signed.
+        // We re-sign to a more usable form and then use it directly.
+        union {
```
Please avoid using unions here as well since reading from `authenticated_value` while `opaque_value` was the last assigned member is UB in C++.

I've prepared a fix which works on my side. You are welcome to just apply that if you are happy with the fix implementation. See commit a29af825c71d70e83445cd4214f7145642201506 from my branch [ptrauth-unwinding-2025-10-19](https://github.com/kovdan01/llvm-project/commits/ptrauth-unwinding-2025-10-19/)

kovdan01 reviewed (changes requested):
@ojhunt Thanks for tons of fixes! We have one more usage of unions in libunwind.cpp which is technically UB.

I've proposed a fix for this. See commit a29af825c71d70e83445cd4214f7145642201506 from my branch [ptrauth-unwinding-2025-10-19](https://github.com/kovdan01/llvm-project/commits/ptrauth-unwinding-2025-10-19/)

As soon as this is addressed, I have no other objections from my side regarding this PR. Thanks for a great work!

ojhunt committed:
Remove union, improve comments.

kovdan01 reviewed (approved):
LGTM, thanks!

ojhunt commented:
Will fix momentarily. I guess gcc does not have `__has_feature`, so I'll look at how libunwind usually does this. We may just gate on the AARCH_PTRAUTH flag instead.

ojhunt commented:
PR up at https://github.com/llvm/llvm-project/pull/164535
